### PR TITLE
bash command line completion

### PR DESCRIPTION
**Pull request by TheTacoScott**
bash command line completion script now installs via setuptools and has it's permissions set correctly on post install.

version numbers bumped, and note added to bottom of main program

I'm not sure how to immediately source the bash completion script once setup.py is done installing without using setuptools hacks that seem improper. A note could be added to the readme to simply relaunch bash. This would keep the setup.py free of less reputable hacks.

### CODE REVIEW

File: data/bash-completion/s4cmd [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    local cur prev words cword possibleparams
+    command=${COMP_WORDS[1]}
+    prev=${COMP_WORDS[COMP_CWORD-1]}
+    prevprev=${COMP_WORDS[COMP_CWORD-2]}
+    _init_completion || return
+
+    #echo -e "\nCurrent: $cur : Prev: $prev : Words: $words : Cword: $cword : Command: $command\n"
+    if [[ $cword -eq 1 ]]; then
+        #command
+        COMPREPLY=( $( compgen -W 'ls put get cp mv sync del du' -- "$cur" ) )
```
Comment (chouhanyang) [inline data/bash-completion/s4cmd]: Not sure what is your convention here, but can you make sure the spaces after ( and before ) is consistent?
For example, next line doesn't have any spaces.

Same for other places.

File: data/bash-completion/s4cmd [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        COMPREPLY=("${COMPREPLY[@]/%/ }")
+    elif [[ $cword -ge 2 ]]; then
+      params=${COMP_WORDS[@]:2}
+      if [ "$prevprev" == "s3" ]
+      then
+        #get initial bucket names
+        if [[ $cur =~ ^\/\/$ ]]; then
+          buckets=`s4cmd ls 2>/dev/null | grep -Eo "s3://.*" | grep -Eo "//.*"`
```
Comment (chouhanyang) [inline data/bash-completion/s4cmd]: Use $() instead of back quote for consistency?

File: data/bash-completion/s4cmd [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+          COMPREPLY=( $( compgen -W "$buckets" -- "$cur" ) )
+          return 0
+        
+        #help get buckets full names
+        elif [[ $cur =~ ^\/\/[^\/]*$ ]]; then
+            buckets=`s4cmd ls 2>/dev/null | grep -Eo "s3://.*" | grep -Eo "//.*"`
```
Comment (chouhanyang) [inline data/bash-completion/s4cmd]: Use $()

File: data/bash-completion/s4cmd [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            COMPREPLY=( $( compgen -W "$buckets" -- "$cur" ) )
+            return 0
+
+        #get contents of dir 
+        elif [[ $cur =~ ^\/\/.*\/$ ]]; then
+            buckets=`s4cmd ls s3:$cur 2>/dev/null | grep -Eo "s3://.*" | grep -Eo "//.*"`
```
Comment (chouhanyang) [inline data/bash-completion/s4cmd]: Use $()

File: data/bash-completion/s4cmd [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            COMPREPLY=( $( compgen -W "$buckets" -- "$cur" ) )
+            return 0
+  
+        #help get buckets contents
+        elif [[ $cur =~ ^\/\/.*\/.*$ ]]; then
+            checkdir=`dirname "$cur"`
```
Comment (chouhanyang) [inline data/bash-completion/s4cmd]: Use $()

File: data/bash-completion/s4cmd [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            buckets=`s4cmd ls s3:$checkdir 2>/dev/null | grep -Eo "s3://.*" | grep -Eo "//.*"`
```
Comment (chouhanyang) [inline data/bash-completion/s4cmd]: Use $()

File: data/bash-completion/s4cmd [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            COMPREPLY=( $( compgen -W "$buckets" -- "$cur" ) )
+            return 0
+        fi
+      fi
+      case "$command" in
+        ls)
+          COMPREPLY=( $( compgen -W "--recursive --show-directory" -- "$cur" ) )
```
Comment (chouhanyang) [inline data/bash-completion/s4cmd]: Same question here about the spaces.

File: data/bash-completion/s4cmd [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+          COMPREPLY=("${COMPREPLY[@]/%/ }")
+          return 0
+          ;;
+        put|get|sync|cp|mv)
+          COMPREPLY=( $( compgen -W "--recursive --sync-check --force --dry-run" -- "$cur" ) )
```
Comment (chouhanyang) [inline data/bash-completion/s4cmd]: Same question here about the spaces.

File: data/bash-completion/s4cmd [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+          COMPREPLY=("${COMPREPLY[@]/%/ }")
+          return 0
+          ;;
+        del)
+          COMPREPLY=( $( compgen -W "--recursive --dry-run" -- "$cur" ) )
```
Comment (chouhanyang) [inline data/bash-completion/s4cmd]: Same question here about the spaces.

File: data/bash-completion/s4cmd [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+          COMPREPLY=("${COMPREPLY[@]/%/ }")
+          return 0
+          ;;
+        du)
+          COMPREPLY=( $( compgen -W "--recursive" -- "$cur" ) )
```
Comment (chouhanyang) [inline data/bash-completion/s4cmd]: Same question here about the spaces.

File: setup.py [hunk 11]
```diff
@@ -20,15 +20,27 @@
...
 
+import os, stat
 from setuptools import setup, find_packages
+from setuptools.command.install import install as _install
 
 __author__ = "Chou-han Yang"
 __copyright__ = "Copyright 2014 BloomReach, Inc."
 __license__ = "http://www.apache.org/licenses/LICENSE-2.0"
-__version__ = "1.5.22"
+__version__ = "1.5.23"
```
Comment (chouhanyang) [inline setup.py]: You also need to change the version in s4cmd.py

**Comment by chouhanyang** [comment]
Few stylish comments and you also need to bump the version in s4cmd.py. Thanks!

**Comment by TheTacoScott** [comment]
All your comments make sense. I'll clean it up. Thanks. :)

**Pull request closed without merging**